packages/tiny-agents/src/lib/webServer.ts

```diff
@@ -0,0 +1,129 @@
+// Generate a simple UUID v4 without dependencies
+const generateUUID = () => {
+	return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
+		const r = (Math.random() * 16) | 0;
+		const v = c === "x" ? r : (r & 0x3) | 0x8;
+		return v.toString(16);
+	});
+};
+const ChatCompletionInputSchema = z.object({
+	messages: z.array(
+		z.object({
+			role: z.enum(["user", "assistant"]),
+			content: z.string(),
+		})
+	),
+	/// Only allow stream: true
+	stream: z.literal(true),
+});
+function getJsonBody(req: IncomingMessage) {
+	return new Promise((resolve, reject) => {
+		let data = "";
+		req.on("data", (chunk) => (data += chunk));
+		req.on("end", () => {
+			try {
+				resolve(JSON.parse(data));
+			} catch (e) {
+				reject(e);
+			}
+		});
+		req.on("error", reject);
+	});
+}
+class ServerResp extends ServerResponse {
+	error(statusCode: number, reason: string) {
+		this.writeHead(statusCode).end(JSON.stringify({ error: reason }));
+	}
+}
+
+export function startServer(agent: Agent): void {
+	const server = createServer({ ServerResponse: ServerResp }, async (req, res) => {
+		res.setHeader(REQUEST_ID_HEADER, generateUUID());
+		res.setHeader("Content-Type", "application/json");
+		if (req.method === "POST" && req.url === "/v1/chat/completions") {
+			let body: unknown;
+			let requestBody: z.infer<typeof ChatCompletionInputSchema>;
+			try {
+				body = await getJsonBody(req);
+			} catch {
+				return res.error(404, "Invalid JSON");
+			}
+			try {
+				requestBody = ChatCompletionInputSchema.parse(body);
+			} catch (err) {
+				if (err instanceof z.ZodError) {
+					return res.error(404, "Invalid ChatCompletionInput body \n" + JSON.stringify(err));
+				}
+				return res.error(404, "Invalid ChatCompletionInput body");
+			}
+			/// Ok, from now on we will send a SSE (Server-Sent Events) response.
+			res.setHeaders(
+				new Headers({
+					"Content-Type": "text/event-stream",
+					"Cache-Control": "no-cache",
+					Connection: "keep-alive",
+				})
+			);
+
+			/// Prepend the agent's prompt
+			const messages = [
+				{
+					role: "system",
+					content: agent.prompt,
+				},
+				...requestBody.messages,
+			];
+
+			for await (const chunk of agent.run(messages)) {
+				if ("choices" in chunk) {
+					res.write(`data: ${JSON.stringify(chunk)}\n\n`);
+				} else {
+					/// Tool call info
+					/// /!\ We format it as a regular chunk!
+					const chunkToolcallInfo = {
+						choices: [
+							{
+								index: 0,
+								delta: {
+									role: "assistant",
+									content:
+										"<tool_call_info>" +
+										`Tool[${chunk.name}] ${chunk.tool_call_id}\n` +
+										chunk.content +
+										"</tool_call_info>",
+								},
+							},
+						],
+						created: Math.floor(Date.now() / 1000),
+						id: chunk.tool_call_id,
+						model: "",
+						system_fingerprint: "",
+					} satisfies ChatCompletionStreamOutput;
+
+					res.write(`data: ${JSON.stringify(chunkToolcallInfo)}\n\n`);
```

Comment:
This is the interesting part of the PR.

I format the tool call info as a "regular" chunk as if it was content generated by the model itself. 🔥 

And I send it as a SSE chunk.